utils/src/main/java/software/amazon/awssdk/utils/http/SdkHttpUtils.java

```diff
@@ -440,5 +446,72 @@ public static Set<String> parseNonProxyHostsProperty() {
         return Collections.emptySet();
     }
 
+    public static Set<String> parseNonProxyHostsEnvironment() {
+        String nonProxyHosts = ProxyEnvironmentSetting.NO_PROXY.getStringValue().orElse(null);
+        if (nonProxyHosts != null && !isEmpty(nonProxyHosts)) {
+            return Arrays.stream(nonProxyHosts.split(","))
+                         .map(String::toLowerCase)
+                         .map(s -> StringUtils.replace(s, "*", ".*?"))
+                         .collect(Collectors.toSet());
+        }
+        return Collections.emptySet();
+    }
+
+    public static Optional<URL> fetchProxyFromEnvironment(String scheme) {
+        Optional<String> environment;
+        if (Objects.equals(scheme, HTTPS)) {
+            environment = ProxyEnvironmentSetting.HTTPS_PROXY.getStringValue();
+        } else if (Objects.equals(scheme, HTTP)) {
+            environment = ProxyEnvironmentSetting.HTTP_PROXY.getStringValue();
+        } else {
+            environment = Optional.empty();
+        }
 
+        return environment.flatMap(value -> {
+            try {
+                URL asUrl = new URL(value);
+                return Optional.of(asUrl);
+            } catch (MalformedURLException cause) {
+                return Optional.empty();
+            }
+        });
+    }
+
+    public static Optional<String> parseUsernameFromUrl(URL value) {
+        String[] parts = parseUserInfoParts(value);
+        if (parts == null) {
+            return Optional.empty();
+        }
+        String trimmed = parts[0].trim();
+        if (trimmed.isEmpty()) {
+            return Optional.empty();
+        }
+        return Optional.of(trimmed);
+    }
+
+    public static Optional<String> parsePasswordFromUrl(URL value) {
```

Comment:
Should this handle passwords that are URLEncoded with special characters?  Maybe after grabbing it out of the parts array, pass it through urlDecode?